### PR TITLE
misc updates

### LINKED ISSUE #2074: Problem with hot module reloading & createGlobalStyle
## System:
 - OS: macOS Sierra 10.12.6
 - CPU: x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
 - Memory: 105.95 MB / 16.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 10.1.0 - ~/.nvm/versions/node/v10.1.0/bin/node
 - Yarn: 1.6.0 - /usr/local/bin/yarn
 - npm: 6.1.0 - ~/.nvm/versions/node/v10.1.0/bin/npm
 - Watchman: 4.7.0 - /usr/local/bin/watchman
## npmPackages:
 - styled-components: ^4.0.0-beta.10 => 4.0.0-beta.10 

Hi there, not sure if I would call this a bug but thought it was worth pointing out.

I have started using the `createGlobalStyle` function in v4, but have run into some issues relating to Hot Module Reloading. Whenever a reload is triggered by something outside the `createGlobalStyle` function, all of its comprising css rules are lost. Here is a simple project which demonstrates this:
https://github.com/frednomoon/styled-hot-reload

It seems from this behaviour that its necessary to reload the module containing the global styles every time the page changes. But, of course, this doesn't happen unless you actually change something inside that module.

I'm guessing its probably possible to solve this with some additional config in my bundler, but obviously would be great to avoid things like that.

### LINKED ISSUE #3008: Global styles does not work in React.StrictMode
## Environment

System:
OS: macOS 10.15.1
CPU: (8) x64 Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
Memory: 857.00 MB / 16.00 GB
Shell: 3.2.57 - /bin/bash
Binaries:
Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
Yarn: 1.16.0 - /usr/local/bin/yarn
npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
npmPackages:
styled-components: ^5.0.1 => 5.0.1
Reproduction
https://codesandbox.io/s/wonderful-hill-6glof

## Steps to reproduce

Load the code sandbox
Click the button "to black" and then click again "to white" and notice how the background did not change back to black.

## Expected Behavior

The should change between black and white

## Actual Behavior

The background stays black after the first press on the toggle button. The global CSS (rendered by GlobalStyle in ComponentBlack) is added twice and only one is removed when unmounting when in Strict Mode. This is because to Strict Mode "intentionally double-invoking" some lifecycle methods to determine side effects that could be a problem in the upcoming async mode in React. When concurrent mode becomes available and React might call render function multiple times the side effect of setting global styles on render will cause issues like this.

As far as I can tell `styled-components` should be `React.Strict` compatible and using the `createGlobalStyle` util is the recommended way to do global styling with `styled-components`, so this seems to be a bug to me.

### PR DESCRIPTION
**Pull request by probablyup**
1. pay attention to SC_VERSION when hashing to provide more effective component isolation when multiple SC instances are running on the page

2. Bundle size improvements

3. Fix `keyframes` not receiving nested stylis context, and therefore missing plugins if applicable

Fixes #3008 #3076 #2993 #3163
Probably fixes #2074

### CODE REVIEW

File: packages/styled-components/rollup.config.js [hunk 1]
```diff
@@ -123,7 +131,8 @@ const browserConfig = {
   plugins: configBase.plugins.concat(
     replace({
       __SERVER__: JSON.stringify(false),
-    })
+    }),
+    minifierPlugin
```
Comment (kitten) [inline packages/styled-components/rollup.config.js:135]: If we're minifying all output bundles (which I'm a fan of) I'd love to make sure to also try out Closure Compiler, which is pretty amazing https://github.com/styled-components/styled-components/blob/d220a202ffc590447d32ce3b5b79e61fe939308b/packages/styled-components/rollup.config.js#L111-L137

This is what we're currently doing for `urql` at least, where we try to squeeze out a lot of bytes here and there https://github.com/FormidableLabs/urql/blob/e031f9b887960a8113890777e8a100973068a36c/scripts/rollup/plugins.js#L98-L137

File: packages/styled-components/src/models/ComponentStyle.js [hunk 2]
```diff
@@ -29,7 +29,11 @@ export default class ComponentStyle {
     this.staticRulesId = '';
     this.isStatic = process.env.NODE_ENV === 'production' && isStaticRules(rules);
     this.componentId = componentId;
-    this.baseHash = hash(componentId);
+
+    // SC_VERSION gives us isolation between multiple runtimes on the page at once
+    // this is improved further with use of the babel plugin "namespace" feature
+    this.baseHash = hash(SC_VERSION + componentId);
```
Comment (kitten) [inline packages/styled-components/src/models/ComponentStyle.js:35]: Do you think it's worth running `hash` on a seed value separately and then continue using `phash` here?

File: packages/styled-components/src/utils/error.js [hunk 3]
```diff
@@ -31,8 +31,8 @@ export default function throwStyledComponentsError(
 ) {
   if (process.env.NODE_ENV === 'production') {
     throw new Error(
-      `An error occurred. See https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/utils/errors.md#${code} for more information.${
-        interpolations.length > 0 ? ` Additional arguments: ${interpolations.join(', ')}` : ''
+      `An error occurred. See https://git.io/JUIaE#${code} for more information.${
```
Comment (kitten) [inline packages/styled-components/src/utils/error.js:34]: Nice!

File: packages/styled-components/src/utils/flatten.js [hunk 4]
```diff
@@ -8,45 +8,47 @@ import isStyledComponent from './isStyledComponent';
...
-          rules.push(`${hyphenate(key)}:`, obj[key], ';')
-
-          return rules
-        }
-        rules.push(`${hyphenate(key)}: ${addUnitIfNeeded(key, obj[key])};`);
+  const rules = [];
+  const keys = Object.keys(obj);
+
+  keys.forEach(key => {
+    if (!isFalsish(obj[key])) {
```
Comment (kitten) [inline packages/styled-components/src/utils/flatten.js:23]: - can we get rid of the nested statements?
- why are we returning `rules`?
- should we just use for-in and `hasOwnProperty` here?

**Review by kitten — APPROVED** [review]

File: packages/styled-components/rollup.config.js (hunk 1 above)
Comment (probablyup) [reply packages/styled-components/rollup.config.js:135]: Shaved off 0.4kB!

File: packages/styled-components/src/models/ComponentStyle.js (hunk 2 above)
Comment (probablyup) [reply packages/styled-components/src/models/ComponentStyle.js:35]: Will bench it

File: packages/styled-components/src/utils/flatten.js (hunk 4 above)
Comment (probablyup) [reply packages/styled-components/src/utils/flatten.js]: This actually made a massive difference in the performance of our object style implementation! Awesome review

File: packages/styled-components/src/models/ComponentStyle.js (hunk 2 above)
Comment (probablyup) [reply packages/styled-components/src/models/ComponentStyle.js]: Minor but noticeable improvement 👍

File: packages/styled-components/rollup.config.js (hunk 1 above)
Comment (probablyup) [reply packages/styled-components/rollup.config.js:135]: So although it reduced the bundle size, something that terser doing actually regressed perf here. Going to hold off on it for now.

**Pull request merged**